### PR TITLE
YPP backend

### PR DESCRIPTION
**Pull request by zeeshanakram3**
implements 
- #29
- #28
- https://github.com/Joystream/marketing/issues/283
- #35

# YoutubeSync

## Required Stack

- Docker
- aws cli

Enter `localhost:3001/docs` to view all api endpoints and their descriptions

For user verification flow only two endpoints are needed

`/users/verify`

`/users/verify-and-save`

## Running the app

1. Install dependencies
   `yarn`

2. Deploy all the necessary infrastructure
```
cd packages/infrastructue
./deploy.sh
```

  This script deploys the relevant dynamoDB tables for needed for the verifications flow.

3. Start the YPP backend server
   `yarn nx run api:serve`

This will start the YPP backend service at localhost:3001

4. Start the frontend app
   `yarn nx run app:serve`
   
Open the browser window at localhost:4200 to interact with the app. Click `VERIFY YOUR YOUTUBE CHANNEL` button to start the verification flow. If verification is successful the YPP will respond back with `email` and `userId` of the verified user. (See response type for `/users/verify` endpoint at `localhost:3001/docs`).  

After that the frontend app need to prepare the request for save the verified channel to the YPP backend state. There is no UI for that so, you can use any api client(e.g. Postman) for making request to `users/verify-and-save` with provided values of `userId`, `joystreamChannelId`, `email`, and optional `referrerId`

### CODE REVIEW

**Review by kdembler — CHANGES_REQUESTED** [review]
Quick first round of feedback after testing:
### Error handling
I think we should introduce some kind of `errorCode` to make it easier to distinct between different error types. Currently if some validation is failed `error` is always `Bad request`, we could repurpose that to create some list of unique error codes to map to specific errors. 

For example, if my channel doesn't meet some criteria I will see:
```json
{
  "statusCode": 400,
  "message": "Channel UCCs9_brR2ADhDVuhrVTjKfQ with 0 subscribers does not \n          meet Youtube Partner Program requirement of 100 subscribers",
  "error": "Bad Request"
}
```
This is problematic because:
a) to understand what kind of error happened we need to do message string matching which can be risky if error message changes in the future
b) we need to parse the message to understand what is the requirement (100 subscribers)

Another issue is that when verifying the eligibility, code will return error on first error found. However, to properly present the status to the user we actually need to see all the failing criteria, not just the first one.

I would suggest error format along those lines:
```json
{
  "errors": [
    {
      "errorCode": "CHANNEL_CRITERIA_UNMET_SUBSCRIBERS",
      "message": "Channel UCCs9_brR2ADhDVuhrVTjKfQ with 0 subscribers does not meet Youtube Partner Program requirement of 100 subscribers",
      "subscribers": 0,
      "requiredSubscribers": 100
    },
    {
        "errorCode": "CHANNEL_CRITERIA_UNMET_VIDEOS",
        "message": "Channel UCCs9_brR2ADhDVuhrVTjKfQ does not meet Youtube Partner Program requirement of at least 10 videos, each 1 month old",
        "videos": 3,
        "requiredVideos": 10
    }
  ]
}
```
Let me know how you think feasible this is.

### Issues

1. When I query `/users/{userId}/videos` I get 500 server error
2. When I query `/channels/{id}`, what is the ID that should be passed here? Seems whatever I pass (I tried user ID, joystream channel ID) I get back
```json
{
  "statusCode": 400,
  "message": "The number of conditions on the keys is invalid"
}
```
3. When I query `/users` I get 500 server error

File: packages/ytube/src/lib/youtubeClient.ts [hunk 1]
```diff
@@ -1,202 +1,266 @@
...
+          meet Youtube Partner Program requirement of ${MINIMUM_SUBSCRIBERS_COUNT} subscribers`
+      )
+    }
+
+    if (channel.statistics.videoCount < MINiMUM_VIDEO_COUNT) {
+      throw new Error(
+        `Channel ${channel.id} with ${channel.statistics.videoCount} videos does not 
+          meet Youtube Partner Program requirement of ${MINiMUM_VIDEO_COUNT} videos`
+      )
+    }
```
[the remark this answers is not in the document]
Comment (zeeshanakram3) [reply packages/ytube/src/lib/youtubeClient.ts:89]: Ah, missed that part. Thanks for pointing this out.

File: packages/ytube/src/lib/youtubeClient.ts [hunk 2]
```diff
@@ -1,202 +1,293 @@
...
+          `Channel ${channel.id} with ${channel.statistics.subscriberCount} subscribers does not ` +
+          `meet Youtube Partner Program requirement of ${MINIMUM_SUBSCRIBERS_COUNT} subscribers`,
+        result: channel.statistics.subscriberCount,
+        expected: MINIMUM_SUBSCRIBERS_COUNT,
+      })
+    }
+
+    // at least MINiMUM_VIDEO_COUNT videos should be one month old
+    const oneMonthsAgo = new Date()
+    oneMonthsAgo.setMonth(oneMonthsAgo.getMonth() - 3)
```
Comment (kdembler) [inline packages/ytube/src/lib/youtubeClient.ts:92]: This talks about one month old but you actually subtract 3 months. Maybe we store both this (MINIMUM_VIDEO_AGE_MONTHS=1) and MINIMUM_CHANNEL_AGE_MONTHS=3 below as const variables up top for better visibility?

**Comment by kdembler** [comment]
Also noticed that if I call `verify` (I will get my YT email back) and then call `verify-and-save` with different email, the different email is not saved and my user still has the YT email attached when I query `/users`

**Comment by zeeshanakram3** [comment]
> The API still fails to start if it fails to connect to RPC endpoint

Addressed. Now API won't crash on Node version > 14

> Also noticed that if I call verify (I will get my YT email back) and then call verify-and-save with different email, the different email is not saved and my user still has the YT email attached when I query /users

Fixed.

File: packages/ytube/src/lib/youtubeClient.ts [hunk 3]
```diff
@@ -1,202 +1,296 @@
...
+import ytdl from 'ytdl-core'
+import Schema$PlaylistItem = youtube_v3.Schema$PlaylistItem
+import Schema$Channel = youtube_v3.Schema$Channel
+import { Channel, ChannelVerificationFailed, ExitCodes, User, Video } from '@youtube-sync/domain'
+import { Readable } from 'stream'
+import { statsRepository } from '..'
+
+// YPP induction criteria, each channel should meet following criteria
+const MINIMUM_SUBSCRIBERS_COUNT = 50
+const MINiMUM_VIDEO_COUNT = 10
```
Comment (kdembler) [inline packages/ytube/src/lib/youtubeClient.ts:12]: Typo, should be `MINIMUM`, all caps

File: packages/ytube/src/lib/youtubeClient.ts [hunk 4]
```diff
@@ -1,202 +1,296 @@
...
+    const videos = (await this.getVideos(channel, MINiMUM_VIDEO_COUNT)).filter(
+      (v) => new Date(v.publishedAt) < oneMonthsAgo
+    )
+    if (videos.length < MINiMUM_VIDEO_COUNT) {
+      errors.push({
+        errorCode: ExitCodes.CHANNEL_CRITERIA_UNMET_VIDEOS,
+        message:
+          `Channel ${channel.id} with ${channel.statistics.videoCount} videos does not meet Youtube ` +
+          `Partner Program requirement of at least ${MINiMUM_VIDEO_COUNT} videos, each 1 month old`,
+        result: channel.statistics.videoCount,
```
Comment (kdembler) [inline packages/ytube/src/lib/youtubeClient.ts:106]: I think this should return `videos.length` instead, right?

File: packages/ytube/src/lib/youtubeClient.ts [hunk 5]
```diff
@@ -1,202 +1,296 @@
...
+    // filter all videos that are older than one month
+    const videos = (await this.getVideos(channel, MINiMUM_VIDEO_COUNT)).filter(
+      (v) => new Date(v.publishedAt) < oneMonthsAgo
+    )
+    if (videos.length < MINiMUM_VIDEO_COUNT) {
+      errors.push({
+        errorCode: ExitCodes.CHANNEL_CRITERIA_UNMET_VIDEOS,
+        message:
+          `Channel ${channel.id} with ${channel.statistics.videoCount} videos does not meet Youtube ` +
+          `Partner Program requirement of at least ${MINiMUM_VIDEO_COUNT} videos, each 1 month old`,
```
Comment (kdembler) [inline packages/ytube/src/lib/youtubeClient.ts:105]: `each 1 month old` should use `MINIMUM_VIDEO_AGE_MONTHS`

File: packages/ytube/src/lib/youtubeClient.ts [hunk 6]
```diff
@@ -1,202 +1,296 @@
...
+
+    // Channel should be at least 3 months old
+    const threeMonthsAgo = new Date()
+    threeMonthsAgo.setMonth(threeMonthsAgo.getMonth() - MINIMUM_CHANNEL_AGE_MONTHS)
+    if (new Date(channel.publishedAt) > threeMonthsAgo) {
+      errors.push({
+        errorCode: ExitCodes.CHANNEL_CRITERIA_UNMET_CREATION_DATE,
+        message:
+          `Channel ${channel.id} with creation time of ${channel.publishedAt} does not ` +
+          `meet Youtube Partner Program requirement of channel being at least 3 months old`,
```
Comment (kdembler) [inline packages/ytube/src/lib/youtubeClient.ts:119]: `at least 3 months old` should use `MINIMUM_CHANNEL_AGE_MONTHS`

**Review by kdembler — CHANGES_REQUESTED** [review]
Additional points:

1. Seems this wasn't resolved? I still see the same behavior
> > Also noticed that if I call verify (I will get my YT email back) and then call verify-and-save with different email, the different email is not saved and my user still has the YT email attached when I query /users

> Fixed.
2. We discussed changing the endpoints 👇 Did you decide to skip this after all?
> /users/verify would be just POST to /users as this call creates a user (it also saves some data, not just verifies)
/users/verify-and-save would be POST to /channels as this call creates an entry in /channels list

**Comment by zeeshanakram3** [comment]
> Seems this wasn't resolved? I still see the same behavior

Sorry, there was an issue in the previous fix. Now resolved

> We discussed changing the endpoints 👇 Did you decide to skip this after all?

Overlooked that before. Now resolved, now `verify-and-save` is POST to `/channels`

File: packages/ytube/src/lib/youtubeClient.ts (hunk 6 above)
Comment (zeeshanakram3) [reply packages/ytube/src/lib/youtubeClient.ts]: addressed in [5c475a3](https://github.com/Joystream/youtube-synch/pull/40/commits/5c475a30e429375f666b84d4b368e5387e68ee42)

File: packages/ytube/src/lib/youtubeClient.ts (hunk 5 above)
Comment (zeeshanakram3) [reply packages/ytube/src/lib/youtubeClient.ts]: addressed in [5c475a3](https://github.com/Joystream/youtube-synch/pull/40/commits/5c475a30e429375f666b84d4b368e5387e68ee42)

File: packages/ytube/src/lib/youtubeClient.ts (hunk 4 above)
Comment (zeeshanakram3) [reply packages/ytube/src/lib/youtubeClient.ts:106]: addressed in [5c475a3](https://github.com/Joystream/youtube-synch/pull/40/commits/5c475a30e429375f666b84d4b368e5387e68ee42)

File: packages/ytube/src/lib/youtubeClient.ts (hunk 3 above)
Comment (zeeshanakram3) [reply packages/ytube/src/lib/youtubeClient.ts]: addressed in [5c475a3](https://github.com/Joystream/youtube-synch/pull/40/commits/5c475a30e429375f666b84d4b368e5387e68ee42)

**Review by kdembler — APPROVED** [review]
Looks good to me, great job!

**Pull request merged**